### PR TITLE
New API to support resolvers as services

### LINKED ISSUE #78: GraphQL: Execute simple queries
We use the [GraphQL parser by Hasura](https://github.com/hasura/graphql-parser-hs)

- [ ] Turn simple `ExecutableDefinition`'s into typed `ServiceQuery`'s.
- [x] Run typed `ServiceQuery`'s against a resolver.

### PR DESCRIPTION
**Pull request by serras**
This is needed to be able to represent GraphQL services within our framework. Essentially, before we have one service per type, and now we have a "package" which may include one or more services.

Closes #78

### CODE REVIEW

**Comment by kutyel** [comment]
Does this PR close any of the open, `graphql` tagged issues? 🤔

**Comment by serras** [comment]
It fixes #77, although we need to rethink what "generic resolver" means in this context.

File: core/rpc/src/Mu/Rpc.hs [hunk 1]
```diff
@@ -27,66 +28,82 @@ import qualified Language.Haskell.TH as TH
...
+type Package' = Package Symbol Symbol
 -- | Services whose names are given by type-level strings.
 type Service' = Service Symbol Symbol
 -- | Annotations for services. At this moment, such
 --   annotations can be of any type.
 type ServiceAnnotation = Type
 
+-- | A package is a set of services.
+data Package serviceName methodName
+  = Package (Maybe serviceName)
```
Comment (kutyel) [inline core/rpc/src/Mu/Rpc.hs]: Why packages "maybe" have a name? 🤔

File: core/rpc/src/Mu/Rpc/Examples.hs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE TupleSections #-}
```
Comment (kutyel) [inline core/rpc/src/Mu/Rpc/Examples.hs]: Lowercase here 😝

File: core/rpc/src/Mu/Rpc/Examples.hs [hunk 3]
```diff
@@ -87,3 +89,40 @@ quickstartServer
...
+apolloServer :: forall m. (MonadServer m)
+             => ServerT Maybe '[ "Book"   ':-> (String, Integer)
+                               , "Author" ':-> Integer ]
+                        ApolloService m _
+apolloServer
+  = Services $     (return . fst :<||>: return . snd :<||>: H0)
+             :<&>: (authorName :<||>: authorBooks :<||>: H0)
+             :<&>: S0
+  where authorName :: Integer -> m String
+        authorName _ = return "alex"  -- this would run in the DB
```
Comment (kutyel) [inline core/rpc/src/Mu/Rpc/Examples.hs]: Use `pure` instead of `return`, to be consistent with our style guide 😉

File: core/rpc/src/Mu/Server.hs [hunk 4]
```diff
@@ -87,11 +99,40 @@ data ServerErrorCode
...
-infixr 5 :<|>:
+data ServicesT (w :: Type -> Type)
+               (chn :: ServiceChain snm) (s :: [Service snm mnm])
+               (m :: Type -> Type) (hs :: [[Type]]) where
+  S0 :: ServicesT w chn '[] m '[]
+  (:<&>:) :: HandlersT w chn (MappingRight chn sname) methods m hs
+          -> ServicesT w chn rest m hss
+          -> ServicesT w chn ('Service sname anns methods ': rest) m (hs ': hss)
+
+infixr 4 :<||>:
```
Comment (kutyel) [inline core/rpc/src/Mu/Server.hs]: Beautiful 😍

File: core/rpc/src/Mu/Server.hs [hunk 5]
```diff
@@ -111,36 +152,59 @@ infixr 5 :<|>:
...
+  (:<||>:) :: Handles w chn args ret m h
+           => (inh -> h) -> HandlersT w chn inh ms m hs
+           -> HandlersT w chn inh ('Method name anns args ret ': ms) m (h ': hs)
+
+infixr 4 :<|>:
+pattern (:<|>:) :: (Handles w chn args ret m h)
+                => h -> HandlersT w chn () ms m hs
+                -> HandlersT w chn () ('Method name anns args ret ': ms) m (h ': hs)
+pattern x :<|>: xs <- (($ ()) -> x) :<||>: xs where
+  x :<|>: xs = noContext x :<||>: xs
```
Comment (kutyel) [inline core/rpc/src/Mu/Server.hs]: Awesome magic ✨

File: core/rpc/src/Mu/Rpc.hs (hunk 1 above)
Comment (serras) [reply core/rpc/src/Mu/Rpc.hs]: Either `language-avro` or `language-protobuf` have a `Maybe Name`, so I changed it here as a consequence.

File: core/rpc/src/Mu/Rpc/Examples.hs (hunk 2 above)
Comment (serras) [reply core/rpc/src/Mu/Rpc/Examples.hs]: Aaargh, ok, I need that automatic formatter you've talked me about

File: core/rpc/src/Mu/Server.hs (hunk 5 above)
Comment (serras) [reply core/rpc/src/Mu/Server.hs]: Crazy magic

File: graphql/mu-graphql.cabal [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  build-depends:       base >=4.12 && <5
+                     , aeson
+                     , mtl
+                     , sop-core
+                     , mu-schema
+                     , mu-rpc
+                     , graphql-parser
+  hs-source-dirs:      src
+  default-language:    Haskell2010
+  ghc-options:         -Wall -fprint-potential-instances
```
Comment (kutyel) [inline graphql/mu-graphql.cabal]: We'll run `cabal-fmt` later on this file if that's ok for you 😉

File: grpc/client/mu-grpc-client.cabal [hunk 7]
```diff
@@ -25,6 +25,7 @@ library
                        Mu.GRpc.Client.Examples
   other-modules:       Mu.GRpc.Client.Internal
   build-depends:       base >=4.12 && <5
+                     , avro
```
Comment (kutyel) [inline grpc/client/mu-grpc-client.cabal]:
```suggestion
                     , avro >= 0.4.7
```

Comment (serras) [reply grpc/client/mu-grpc-client.cabal]: This doesn't belong to this PR.

File: grpc/client/mu-grpc-client.cabal [hunk 8]
```diff
@@ -30,6 +30,7 @@ library
   other-modules:    Mu.GRpc.Client.Internal
   build-depends:
       async
+    , avro
```
Comment (kutyel) [inline grpc/client/mu-grpc-client.cabal]:
```suggestion
    , avro               >=0.4.7
```

File: cabal-fmt.sh [hunk 9]
```diff
@@ -1 +1 @@
-find . -name '*.cabal' -exec sh -c 'cabal-fmt $0 > output.tmp; mv output.tmp $0' {} ';'
+find . -name '*.cabal' -exec sh -c 'cabal-fmt -i $0' {} ';'
```
Comment (kutyel) [inline cabal-fmt.sh]: @rachelcarmena here's the improved version! 😉 thanks a lot for your help! ❤️

Comment (rachelcarmena) [reply cabal-fmt.sh]: Thanks to you for letting me know about it :heart: @serras another :heart: for you, as well :sweat_smile:

File: graphql/src/Mu/GraphQL/Query/Parse.hs [hunk 10]
```diff
@@ -63,9 +63,9 @@ instance
   (KnownSymbol mname, ParseMethod p ms, ParseArgs p args, ParseReturn p r) =>
   ParseMethod p ('Method mname manns args ('RetSingle r) ': ms)
   where
-  selectMethod (GQL.unName -> wanted) args sels
+  selectMethod w@(GQL.unName -> wanted) args sels
```
Comment (kutyel) [inline graphql/src/Mu/GraphQL/Query/Parse.hs]: Beautiful ❤️

**Pull request merged**